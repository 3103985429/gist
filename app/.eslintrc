{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
