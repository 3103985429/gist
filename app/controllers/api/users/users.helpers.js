const async = require('async');

const User = require('../../../models/users');
const Team = require('../../../models/teams');
const mailer = require('../../../libs/utils/nodemailer');
const sample_dataset = require('../sample_dataset');

const isEnterprise = process.env.NODE_ENV === 'enterprise';

/**
 * if team is enterprise, get team object
 * @param {object} req - the req
 * @param {function} callback
 */
const getEnterpriseTeam = (req, callback) => {
    if (isEnterprise) {
        Team.findOne({ subdomain: process.env.SUBDOMAIN }, (err, team) => {
            callback(err, team, req);
        });
    } else {
        callback(null, null, req);
    }
};

/**
 * create a new team if there is no enterprise team
 * @param {object} enterpriseTeam - found enterprise team or null
 * @param {object} req - request
 * @param {function} callback
 */
const newTeam = (enterpriseTeam, req, callback) => {
    const { _id: userId, _team: team } = req.body;

    if (!enterpriseTeam) {
        team.admin = [userId];
        team.superTeam = (team.subdomain === 'schema' || isEnterprise);
        team.isEnterprise = isEnterprise;

        Team.create(team, (err, createdTeam) => {
            if (err) {
                return callback(err);
            } else if (!createdTeam) {
                return callback(new Error('Could not create team'));
            } else {
                if (createdTeam.title !== 'sampleTeam' && (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'staging')) {
                    return sample_dataset.delegateDatasetDuplicationTasks(userId, createdTeam, err => {
                        if (err) {
                            return callback(err);
                        }

                        callback(null, req, createdTeam);
                    });
                }

                callback(null, req, createdTeam);
            }
        });
    } else {
        callback(null, req, enterpriseTeam);
    }
};

/**
 * find and update the user object
 * @param {object} req - request
 * @param {object} team
 * @param {function} callback
 */
const updateUser = (req, team, callback) => {
    const { firstName, lastName, password, _id: userId } = req.body;

    User.findById(userId, (err, user) => {
        if (err || !user) {
            return callback(err || new Error(`Could not find user with id: ${userId}`));
        }

        if (req.isAnonymous && user.signedUp) {
            return callback(new Error('User already registered'));
        } else {
            user.signedUp = true;
        }

        if (user.provider === 'local' && password && (!user.hash || !user.salt)) {
            user.setPassword(password);
        }
        user.firstName = firstName;
        user.lastName = lastName;
        user._team = [team._id];
        user.defaultLoginTeam = team._id;
        callback(null, user, team);
    });
};

/**
 * find user and check if he has already signed in before
 * @param {object} req - request
 * @param {function} callback
 */
const checkUser = (req, callback) => {
    const { _id: userId } = req.body;

    User.findById(userId, (err, user) => {
        if (err || !user) {
            return callback(err || new Error(`Could not find user with id: ${userId}`));
        }

        if (user.signedUp) {
            return callback(new Error('User already registered'));
        }

        return callback(null);
    });
};

/**
 * Set activated to true.
 * Users creating their own team will need to click the email verification link to activate.
 * Invited users have already arrived at the signup page via the email link.
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const activateUser = (user, team, callback) => {
    user.activated = true;
    callback(null, user, team);
};

/**
 * save the updated user
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const saveUpdatedUser = (user, team, callback) => {
    user.save((err, updatedUser) => {
        callback(err, updatedUser, team);
    });
};

/**
 * if not enterprise, call method to send email notifying user that their team was created
 * if user is activated (aka invited user) - do nothing
 * if user is not activated (aka creating team) - call method to send an activation email link
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const sendNotificationEmail = (user, team, callback) => {
    if (!isEnterprise) {
        team.notifyNewTeamCreation();
    }

    if (user.activated) {
        callback(null, user);
    } else {
        mailer.sendActivationEmail(user, err => {
            callback(err, user);
        });
    }
};

/**
 * new user signup process
 */
module.exports.newUser = (req, callback) => {
    async.waterfall([
        async.apply(checkUser, req),
        async.apply(getEnterpriseTeam, req),
        newTeam,
        updateUser,
        saveUpdatedUser,
        sendNotificationEmail,
    ], (err, user) => {
        callback(err, user);
    });
};

/**
 * invited user signup process
 */
module.exports.userInvited = (req, callback) => {
    const { _team: team } = req.body;

    async.waterfall([
        async.apply(checkUser, req),
        async.apply(updateUser, req, team),
        activateUser,
        saveUpdatedUser,
    ], (err, user) => {
        callback(err, user);
    });
};
