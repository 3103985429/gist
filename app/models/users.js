const mongoose = require('mongoose');
const crypto = require('crypto');
const winston = require('winston');
const { isEmpty } = require('lodash');

const { Schema } = mongoose;

const userSchema = new Schema({
    email: {
        type: String,
        unique: true,
        required: true,
    },
    firstName: String,
    lastName: String,
    profileImageUrl: String,
    provider: String,
    hash: String,
    salt: String,
    loginAttempts: { type: Number, required: true, default: 0 },
    lockUntil: { type: Number, default: 1 },
    activated: {
        type: Boolean,
        default: false,
    },
    sampleImported: {
        type: Boolean,
        default: false,
    },
    _team: [{ type: Schema.Types.ObjectId, ref: 'Team' }],
    active: {
        type: Boolean,
        default: true,
    },
    inviteLink: String,
    // denotes those users with non-schema emails as superAdmins
    superUser: Boolean,

    invited: Object, //temp object to store invited new user
    canCreateNewViz: Array,
    canCreateNewArticle: Array,
    canCreateNewSite: Array,
    _editors: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _viewers: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _articleEditors: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _articleViewers: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _siteEditors: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    _siteViewers: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    defaultLoginTeam: { type: Schema.Types.ObjectId, ref: 'Team' },
    inviteToken: { type: Schema.Types.ObjectId, ref: 'Token' },
    confirmation_code_hash: Object,
    signedUp: {
        type: Boolean,
        default: false,
    },
    inviter: String,
    role: String,
    smartsheetToken: Object, // Smartsheet token object to hold smartsheet token connection information
    pipedriveToken: Object, // Pipedrive token object
    datadotworldToken: Object, // Data.world token object
    salesforceToken: Object, // Salesforce token object
}, { timestamps: true });

const deepPopulate = require('mongoose-deep-populate')(mongoose);
userSchema.plugin(deepPopulate, { whitelist: ['defaultLoginTeam.datasourceDescriptions', '_team.datasourceDescriptions'] });

userSchema.methods.setPassword = function (password) {
    this.salt = crypto.randomBytes(16).toString('hex');
    this.hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
};

userSchema.methods.validPassword = function (password) {
    const hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
    return this.hash === hash;
};

userSchema.methods.isSuperAdmin = function () {
    const re = new RegExp('\\+superuser@schemadesign.com$');
    return re.test(this.email) || this.superUser === true;
};

userSchema.methods.isAdmin = function (team) {
    if (!team) {
        winston.warn('isAdmin was called without team param');
        return false;
    }

    if (isEmpty(team.admin)) {
        winston.warn('isAdmin was called on team without admin set');
        return false;
    }

    const userId = this._id.toString();

    return team.admin.some(id => id.toString() === userId);
};

userSchema.methods.isTeamMember = function (team) {
    if (!team) {
        winston.warn('isTeamMember was called without team param');
        return false;
    }
    return this._teams.some(team => team.equals(team._id));
};

userSchema.methods.isLocked = function () {
    // check for a future lockUntil timestamp
    return this.lockUntil && this.lockUntil > Date.now();
};

userSchema.methods.checkLoginAttempts = function () {
    // Add properties to prevent brute force login if user doesn't have them
    if (!this.lockUntil) {
        this.lockUntil = 1;
    }
    if (!this.loginAttempts) {
        this.loginAttempts = 0;
    }
};

userSchema.methods.lockoutUser = function () {
    this.loginAttempts += 1;
    if (this.loginAttempts > 5) { // Lock user out after 6 failed login attempts
        this.lockUntil = Date.now() + (24 * 60 * 60 * 1000); // Lock user out for 24 hours
    }
};

userSchema.methods.resetLock = function () {
    this.loginAttempts = 0;
    this.lockUntil = 1;
};

module.exports = mongoose.model('User', userSchema);
