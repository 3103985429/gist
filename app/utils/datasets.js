const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
