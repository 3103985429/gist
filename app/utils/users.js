const Users = require('../models/users');

module.exports = {
    getUser,
};

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getUser(id) {
    return await Users
        .findById(id)
        .lean()
        .exec();
}
