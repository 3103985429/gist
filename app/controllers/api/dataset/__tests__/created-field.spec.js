import request from 'supertest';

import { hydrateDb, loginAndGetSessionCookie } from '../../../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../../../internals/testing/backend/fixtures/datasets';
import { user1 } from '../../../../../internals/testing/backend/fixtures/users';
import app from '../../../../../app';

const { Lazy_Shared_ProcessedRowObject_MongooseContext } = require('../../../../models/processed_row_objects');

const getProcessedRowObjectsModel = (datasetId) => Lazy_Shared_ProcessedRowObject_MongooseContext(datasetId).Model;

describe('Created field', () => {
    let appRequest, createResponse, sessionCookie;
    const ProcessedRowObjects = getProcessedRowObjectsModel(pokemon1._id);
    const newField = 'new-field';

    beforeEach(async () => {
        await hydrateDb();

        appRequest = request(app);
        sessionCookie = await loginAndGetSessionCookie(appRequest, user1);
        createResponse = await appRequest
            .post(`/api/dataset/content/${pokemon1._id}/createField/${newField}`)
            .send({ dataType: 'Text' })
            .set('Cookie', sessionCookie)
            .set('Host', `glitter.${process.env.HOST}`);
    });

    it('should create a new field', async () => {
        const data = await ProcessedRowObjects.find({ [`rowParams.${newField}`]: { $exists: true } }, { rowParams: 1 });

        expect(data).toMatchSnapshot();
        expect(createResponse).toMatchSnapshot();
    });

    it('should remove a field', async () => {
        const removeResponse = await appRequest
            .post(`/api/dataset/content/${pokemon1._id}/removeField/new-field`)
            .send({ dataType: 'Text' })
            .set('Cookie', sessionCookie)
            .set('Host', `glitter.${process.env.HOST}`);
        const data = await ProcessedRowObjects.find({ [`rowParams.${newField}`]: { $exists: true } });

        expect(data).toMatchSnapshot();
        expect(removeResponse).toMatchSnapshot();
    });
});
