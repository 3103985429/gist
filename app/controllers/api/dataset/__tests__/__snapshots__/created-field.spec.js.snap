// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Created field should create a new field 1`] = `
Array [
  Object {
    "_id": "5b05346ad7a86407e5ef783d",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/2/21/001Bulbasaur.png/250px-001Bulbasaur.png",
      "Max CP": 1079,
      "Max HP": 83,
      "Name": "Bulbasaur",
      "Pokemon No_": 1,
      "Type 1": "Grass",
      "Type 2": "Poison",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef783e",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/73/002Ivysaur.png/250px-002Ivysaur.png",
      "Max CP": 1643,
      "Max HP": 107,
      "Name": "Ivysaur",
      "Pokemon No_": 2,
      "Type 1": "Grass",
      "Type 2": "Poison",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef783f",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/a/ae/003Venusaur.png/250px-003Venusaur.png",
      "Max CP": 2598,
      "Max HP": 138,
      "Name": "Venusaur",
      "Pokemon No_": 3,
      "Type 1": "Grass",
      "Type 2": "Poison",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7840",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/73/004Charmander.png/250px-004Charmander.png",
      "Max CP": 962,
      "Max HP": 73,
      "Name": "Charmander",
      "Pokemon No_": 4,
      "Type 1": "Fire",
      "Type 2": "None",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7841",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/4/4a/005Charmeleon.png/250px-005Charmeleon.png",
      "Max CP": 1568,
      "Max HP": 103,
      "Name": "Charmeleon",
      "Pokemon No_": 5,
      "Type 1": "Fire",
      "Type 2": "None",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7842",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/7/7e/006Charizard.png/250px-006Charizard.png",
      "Max CP": 2620,
      "Max HP": 135,
      "Name": "Charizard",
      "Pokemon No_": 6,
      "Type 1": "Fire",
      "Type 2": "Flying",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7843",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/3/39/007Squirtle.png/250px-007Squirtle.png",
      "Max CP": 1015,
      "Max HP": 81,
      "Name": "Squirtle",
      "Pokemon No_": 7,
      "Type 1": "Water",
      "Type 2": "None",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7844",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/0/0c/008Wartortle.png/250px-008Wartortle.png",
      "Max CP": 1594,
      "Max HP": 105,
      "Name": "Wartortle",
      "Pokemon No_": 8,
      "Type 1": "Water",
      "Type 2": "None",
      "new-field": "",
    },
  },
  Object {
    "_id": "5b05346ad7a86407e5ef7845",
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/0/02/009Blastoise.png/250px-009Blastoise.png",
      "Max CP": 2560,
      "Max HP": 137,
      "Name": "Blastoise",
      "Pokemon No_": 9,
      "Type 1": "Water",
      "Type 2": "None",
      "new-field": "",
    },
  },
]
`;

exports[`Created field should create a new field 2`] = `
Object {
  "body": Object {
    "JSONPath": "undefined",
    "_otherSources": Array [],
    "_team": Object {
      "admin": Array [
        "5a42bf07d3bc0b20648290e0",
      ],
      "brandColor": Object {
        "label": "#2A2A2A",
        "select": "#2A2A2A",
      },
      "colorPalette": Array [
        "#1B6DFC",
        "#0CB7FA",
        "#FA79FC",
        "#FB3705",
        "#F9D307",
        "#FA9007",
        "#7D5807",
        "#B2E606",
        "#069908",
        "#7B07FD",
      ],
      "datasourceDescriptions": Array [
        "5a42c26e303770207ce8f83b",
        "5b033d64ef29594d18bb18aa",
      ],
      "font": "calibre",
      "hasEnterpriseLicense": false,
      "pages": Array [
        "5ac26b456f8073766713bcee",
        "5ac26b3d6f8073766713bced",
        "5ac26b506f8073766713bcef",
        "5ac26b636f8073766713bcf0",
        "aaa26b456f8073766713bcee",
      ],
      "sites": Array [],
      "subdomain": "glitter",
      "superTeam": true,
      "title": "Glitter",
    },
    "author": "5a42bf07d3bc0b20648290e0",
    "brandColor": Object {
      "accent": "#005CB5",
    },
    "clones": 0,
    "columns": Array [
      Object {
        "data_type": "Text",
        "name": "Image URL",
        "operation": "ToString",
      },
      Object {
        "data_type": "Number",
        "name": "Max HP",
        "operation": "ToInteger",
      },
      Object {
        "data_type": "Number",
        "name": "Max CP",
        "operation": "ToInteger",
      },
      Object {
        "data_type": "Text",
        "name": "Type 2",
        "operation": "ToString",
      },
      Object {
        "data_type": "Text",
        "name": "Type 1",
        "operation": "ToString",
      },
      Object {
        "data_type": "Text",
        "name": "Name",
        "operation": "ToString",
      },
      Object {
        "data_type": "Number",
        "name": "Pokemon No_",
        "operation": "ToInteger",
      },
      Object {
        "createdField": true,
        "data_type": "Text",
        "name": "new-field",
        "operation": "ToString",
        "sample": "",
      },
    ],
    "contentEdited": true,
    "createdFields": Array [
      Object {
        "createdField": true,
        "data_type": "Text",
        "name": "new-field",
        "operation": "ToString",
        "sample": "",
      },
    ],
    "customFieldsToProcess": Array [],
    "dirty": 0,
    "fe_displayTitleOverrides": Object {
      "Type 1": "Type1_overridden",
    },
    "fe_excludeFields": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
      "new-field": false,
    },
    "fe_excludeFieldsObjDetail": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
      "new-field": false,
    },
    "fe_fieldDisplayOrder": Array [
      "new-field",
    ],
    "fe_filters": Object {
      "default": Object {},
      "excludeFields": Array [],
      "fabricated": Array [],
      "fieldsAsRangeSlider": Array [],
      "fieldsCommaSeparatedAsIndividual": Array [],
      "fieldsMultiSelectable": Array [],
      "fieldsNotAvailable": Array [
        "new-field",
      ],
      "fieldsSortableByInteger": Array [],
      "fieldsSortableInReverseOrder": Array [],
      "keywords": Array [],
      "valuesToExcludeByOriginalKey": Object {
        "_all": Array [],
      },
    },
    "fe_image": Object {
      "detailView": true,
      "field": "",
      "gallery": true,
      "overwrite": false,
      "scraped": false,
      "timeline": true,
    },
    "fe_listed": true,
    "fe_nestedObject": Object {
      "criteria": Object {},
      "fields": Array [],
    },
    "fe_viewOptions": Object {
      "enableAccessibility": false,
      "enableLastUpdated": false,
      "filtersRail": true,
      "fullscreenExpand": true,
      "insightsExplorer": true,
      "insightsExplorerEmbed": true,
      "search": true,
      "viewControls": true,
      "viewInteractivity": true,
    },
    "fe_views": Object {
      "default_view": "gallery",
      "views": Object {
        "gallery": Object {
          "defaultSortByColumnName": "Image URL",
          "visible": true,
        },
        "pieChart": Object {
          "defaultGroupByColumnName": "Name",
          "visible": true,
        },
        "pieSet": Object {
          "defaultAggregateByColumnName": "Max HP",
          "defaultChartByColumnName": "Type 1",
          "defaultGroupByColumnName": "Type 2",
          "fieldsNotAvailableAsAggregateByColumns": Array [],
          "fieldsNotAvailableAsChartByColumns": Array [],
          "fieldsNotAvailableAsGroupByColumns": Array [],
          "valuesToExcludeByOriginalKey": Array [],
          "visible": true,
        },
        "timeline": Object {
          "visible": true,
        },
        "treemap": Object {
          "defaultAggregateByColumnName": "Max HP",
          "defaultChartByColumnName": "Name",
          "defaultGroupByColumnName": "Type 1",
          "fieldsNotAvailableAsAggregateByColumns": Array [],
          "fieldsNotAvailableAsChartByColumns": Array [],
          "fieldsNotAvailableAsGroupByColumns": Array [],
          "visible": true,
        },
      },
    },
    "fe_visible": true,
    "fileName": "pokemonGO.csv",
    "firstImport": 0,
    "format": "CSV",
    "importRevision": 1,
    "imported": true,
    "includeEmptyFields": true,
    "isPublic": true,
    "jobId": 0,
    "lastImportInitiatedBy": null,
    "lastImportTriggeredBy": "manual",
    "objectTitle": "Pokemon No_",
    "openDownload": false,
    "preserveEditedData": true,
    "previous_datasets": Array [],
    "raw_rowObjects_coercionScheme": Object {
      "Image URL": Object {
        "operation": "ToString",
      },
      "Max CP": Object {
        "operation": "ToInteger",
      },
      "Max HP": Object {
        "operation": "ToInteger",
      },
      "Name": Object {
        "operation": "ToString",
      },
      "Pokemon No_": Object {
        "operation": "ToInteger",
      },
      "Type 1": Object {
        "operation": "ToString",
      },
      "Type 2": Object {
        "operation": "ToString",
      },
      "new-field": Object {
        "operation": "ToString",
      },
    },
    "relationshipFields": Array [],
    "replaced": false,
    "replacement": false,
    "sample": false,
    "skipImageScraping": false,
    "state": "approved",
    "tabDestination": 0,
    "title": "Pokemon",
    "uid": "pokemon",
    "updatedBy": "5a42bf07d3bc0b20648290e0",
    "updatedContent": Object {
      "created": Object {},
      "edited": Object {},
      "new": true,
      "publishedStatus": Object {},
    },
    "urls": Array [],
  },
  "statusCode": 200,
}
`;

exports[`Created field should remove a field 1`] = `Array []`;

exports[`Created field should remove a field 2`] = `
Object {
  "body": Object {
    "JSONPath": "undefined",
    "_otherSources": Array [],
    "_team": "5a42bf30d3bc0b20648290e1",
    "author": "5a42bf07d3bc0b20648290e0",
    "brandColor": Object {
      "accent": "#005CB5",
    },
    "clones": 0,
    "columns": Array [
      Object {
        "data_type": "Text",
        "name": "Image URL",
        "operation": "ToString",
      },
      Object {
        "data_type": "Number",
        "name": "Max HP",
        "operation": "ToInteger",
      },
      Object {
        "data_type": "Number",
        "name": "Max CP",
        "operation": "ToInteger",
      },
      Object {
        "data_type": "Text",
        "name": "Type 2",
        "operation": "ToString",
      },
      Object {
        "data_type": "Text",
        "name": "Type 1",
        "operation": "ToString",
      },
      Object {
        "data_type": "Text",
        "name": "Name",
        "operation": "ToString",
      },
      Object {
        "data_type": "Number",
        "name": "Pokemon No_",
        "operation": "ToInteger",
      },
    ],
    "contentEdited": true,
    "createdFields": Array [],
    "customFieldsToProcess": Array [],
    "dirty": 0,
    "fe_displayTitleOverrides": Object {
      "Type 1": "Type1_overridden",
    },
    "fe_excludeFields": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
    },
    "fe_excludeFieldsObjDetail": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
    },
    "fe_fieldDisplayOrder": Array [],
    "fe_filters": Object {
      "default": Object {},
      "excludeFields": Array [],
      "fabricated": Array [],
      "fieldsAsRangeSlider": Array [],
      "fieldsCommaSeparatedAsIndividual": Array [],
      "fieldsMultiSelectable": Array [],
      "fieldsNotAvailable": Array [],
      "fieldsSortableByInteger": Array [],
      "fieldsSortableInReverseOrder": Array [],
      "keywords": Array [],
      "valuesToExcludeByOriginalKey": Object {
        "_all": Array [],
      },
    },
    "fe_image": Object {
      "detailView": true,
      "field": "",
      "gallery": true,
      "overwrite": false,
      "scraped": false,
      "timeline": true,
    },
    "fe_listed": true,
    "fe_nestedObject": Object {
      "criteria": Object {},
      "fields": Array [],
    },
    "fe_viewOptions": Object {
      "enableAccessibility": false,
      "enableLastUpdated": false,
      "filtersRail": true,
      "fullscreenExpand": true,
      "insightsExplorer": true,
      "insightsExplorerEmbed": true,
      "search": true,
      "viewControls": true,
      "viewInteractivity": true,
    },
    "fe_views": Object {
      "default_view": "gallery",
      "views": Object {
        "gallery": Object {
          "defaultSortByColumnName": "Image URL",
          "visible": true,
        },
        "pieChart": Object {
          "defaultGroupByColumnName": "Name",
          "visible": true,
        },
        "pieSet": Object {
          "defaultAggregateByColumnName": "Max HP",
          "defaultChartByColumnName": "Type 1",
          "defaultGroupByColumnName": "Type 2",
          "fieldsNotAvailableAsAggregateByColumns": Array [],
          "fieldsNotAvailableAsChartByColumns": Array [],
          "fieldsNotAvailableAsGroupByColumns": Array [],
          "valuesToExcludeByOriginalKey": Array [],
          "visible": true,
        },
        "timeline": Object {
          "visible": true,
        },
        "treemap": Object {
          "defaultAggregateByColumnName": "Max HP",
          "defaultChartByColumnName": "Name",
          "defaultGroupByColumnName": "Type 1",
          "fieldsNotAvailableAsAggregateByColumns": Array [],
          "fieldsNotAvailableAsChartByColumns": Array [],
          "fieldsNotAvailableAsGroupByColumns": Array [],
          "visible": true,
        },
      },
    },
    "fe_visible": true,
    "fileName": "pokemonGO.csv",
    "firstImport": 0,
    "format": "CSV",
    "importRevision": 1,
    "imported": true,
    "includeEmptyFields": true,
    "isPublic": true,
    "jobId": 0,
    "lastImportInitiatedBy": null,
    "lastImportTriggeredBy": "manual",
    "objectTitle": "Pokemon No_",
    "openDownload": false,
    "preserveEditedData": true,
    "previous_datasets": Array [],
    "raw_rowObjects_coercionScheme": Object {
      "Image URL": Object {
        "operation": "ToString",
      },
      "Max CP": Object {
        "operation": "ToInteger",
      },
      "Max HP": Object {
        "operation": "ToInteger",
      },
      "Name": Object {
        "operation": "ToString",
      },
      "Pokemon No_": Object {
        "operation": "ToInteger",
      },
      "Type 1": Object {
        "operation": "ToString",
      },
      "Type 2": Object {
        "operation": "ToString",
      },
    },
    "relationshipFields": Array [],
    "replaced": false,
    "replacement": false,
    "sample": false,
    "skipImageScraping": false,
    "state": "approved",
    "tabDestination": 0,
    "title": "Pokemon",
    "uid": "pokemon",
    "updatedBy": "5a42bf07d3bc0b20648290e0",
    "updatedContent": Object {
      "created": Object {},
      "edited": Object {},
      "new": true,
      "publishedStatus": Object {},
    },
    "urls": Array [],
  },
  "statusCode": 200,
}
`;
